cavern/src/main/java/org/opencadc/cavern/files/PutAction.java

```diff
@@ -183,15 +185,53 @@ public void doAction() throws Exception {
 
             // doing this last because it requires chown which is most likely to fail during experimentation
             log.debug("restore owner & group");
-            UserPrincipal owner = NodeUtil.getOwner(getUpLookupSvc(), node);
-            GroupPrincipal group = NodeUtil.getDefaultGroup(getUpLookupSvc(), owner);
-            NodeUtil.setPosixOwnerGroup(rootPath, target, owner, group);
-            
+            restoreOwnNGroup(rootPath, node);
         } catch (AccessControlException | AccessDeniedException e) {
             log.debug(e);
             syncOutput.setCode(403);
+        } catch (IOException e) {
+            if (e.getMessage().contains("quota exceeded")) {
+                // TODO: Replace the quota limit below with the number of bytes 
+                //       remaining in the quota when we can determine it.
+                // get quota limit from a parent container
+                Path rootPath = Paths.get(getRoot());
+                restoreOwnNGroup(rootPath, node);
+                Node curNode = node.getParent();
+                String limitValue = curNode.getPropertyValue(VOS.PROPERTY_URI_QUOTA);
+                while (limitValue == null) {
+                    curNode = curNode.getParent();
+                    try {
+                        limitValue = curNode.getPropertyValue(VOS.PROPERTY_URI_QUOTA);
+                    } catch (NullPointerException ex) {
+                        // quota limit not defined in a parent container, exit while loop
+                        break;
+                    }
+                }
+
+                // -1 to indicate quota limit not defined
+                long limit = -1;
+                String msg = e.getMessage();
+                if (limitValue == null) {
+                    // VOS.PROPERTY_URI_QUOTA attribute is not set on the node
+                    msg = "VOS.PROPERTY_URI_QUOTA attribute not set, " + e.getMessage();
```

Comment:
This addition to the comment will probably confuse users because it would be a bug in our system.  (quota enforced by not set?). It's good that you're dealing with it though.  I would put a log.warn() message in there instead.

Reply:
Done